src/main/java/tree/ImpTargetTreeNode.kt

```diff
@@ -1,30 +1,71 @@
 package tree
 
-class ImpTargetTreeNode : TargetTreeNode {
-    override fun moveRight(): TargetTreeNode {
-        println("moveRight")
-        return this
+class ImpTargetTreeNode(
+    override val name: String = "",
+    override val tag: String = "",
+    override val id: Int = -1
+) : TargetTreeNode {
+
+    override var children: List<TargetTreeNode> = mutableListOf()
+    private var parent: TargetTreeNode? = null
+
+    override fun moveToRightSibling(): TargetTreeNode? {
+        return (parent as? ImpTargetTreeNode)?.getRightSibling(this)
+    }
+
+    override fun moveToLeftSibling(): TargetTreeNode? {
+        return (parent as? ImpTargetTreeNode)?.getLeftSibling(this)
+    }
+
+    override fun moveToParent(): TargetTreeNode? {
+        return parent
     }
 
-    override fun moveLeft(): TargetTreeNode {
-        println("moveLeft")
-        return this
+    override fun moveToFirstChild(): TargetTreeNode? {
+        if (children.isNotEmpty())
+            return children.first()
+        return null
     }
 
-    override fun moveUp(): TargetTreeNode {
-        println("moveUp")
-        return this
+    override fun toString(): String {
+        var str = "$name:$tag"
+        if (children.isNotEmpty())
+            str += END_LINE_STRING
+        str += children.joinToString(separator = END_LINE_STRING).prependIndent(INDENT_STRING)
+        return str
     }
 
-    override fun moveDown(): TargetTreeNode {
-        println("moveDown")
-        return this
+    fun setParent(newParent: TargetTreeNode?) {
+        parent = newParent
     }
 
-    override val name: String
-        get() = "name"
+    private fun getRightSibling(son: TargetTreeNode): TargetTreeNode? {
+        if (children.isNotEmpty()) {
+            val index = (children as List<ImpTargetTreeNode>).binarySearch(son as ImpTargetTreeNode, ImpTargetTreeNodeComparator())
+            if (index >= 0 && index + 1 < children.size)
+                return children[index + 1]
+        }
+        return null
+    }
+
+    private fun getLeftSibling(son: TargetTreeNode): TargetTreeNode? {
+        if (children.isNotEmpty()) {
+            val index = (children as List<ImpTargetTreeNode>).binarySearch(son as ImpTargetTreeNode, ImpTargetTreeNodeComparator())
+            if (index > 0)
+                return children[index - 1]
+        }
+        return null
+    }
+
+    fun preorder() {
+        println("$name:$tag($id)")
+        for (child in children) {
+            (child as ImpTargetTreeNode).preorder()
+        }
+    }
```

Comment:
This function is never used in this release and was previously used as a testing mechanism but we now have more robust ones, so i think it should disappear.
```suggestion
```

Reply:
No, maybe this function will be used after in the project, is better keep it

Reply:
Sounds right